flightControl/src/AvState.cpp

```diff
@@ -0,0 +1,262 @@
+    else if (data.speed.z <= 0)
+    {
+        return State::DESCENT;
+    }
+    return State::ASCENT;
+}
+
+State AvState::fromCalibration(SensFiltered data, UPLink uplink, bool status)
+{
+    if (error() || uplink.id == CMD_ID::AV_CMD_ABORT)
+    {
+        return State::ERRORGROUND;
+        //check if the calibration is done
+    }
+    else if (status)
+    {
+        return State::MANUAL;
+    }
+    else
+    {
+        return State::CALIBRATION;
+    }
+}
+
+
+State AvState::fromErrorGround(SensFiltered data, UPLink uplink)
+{
+    if (uplink.id == CMD_ID::AV_CMD_RECOVER)
+    {
+        return State::IDLE;
+    }
+}
+
+State AvState::fromErrorFlight()
+{
+    return State::ERRORFLIGHT;
+}
+
+State AvState::fromThrustSequence(SensFiltered data, UPLink uplink)
+{
+    double speed = (data.speed.x * data.speed.x + data.speed.y * data.speed.y + data.speed.z * data.speed.z);
+    if (uplink.id == CMD_ID::AV_CMD_ABORT)
+    {
+        return State::ERRORFLIGHT;
+    }
+    else if (data.speed.z>0) //TODO we need a ignite bool var
+                                //given by the prop board
+    {
+        return State::ASCENT;
+    }
+    else
+    {
+        return State::THRUSTSEQUENCE;
+    }
+}
+
+State AvState::fromManual(SensFiltered data, UPLink uplink)
+{
+    if (error() || uplink.id == CMD_ID::AV_CMD_ABORT)
+    {
+        return State::ERRORGROUND;
+    }
+        //TODO: recheck i threholds are the wanted ones
+    else if (uplink.id == CMD_ID::AV_CMD_ARM &&
+             data.fuel_pressure > thresholds.fuel_pressure_wanted &&
+             data.LOX_pressure > thresholds.lox_pressure_wanted &&
+             data.chamber_pressure > thresholds.chamber_pressure_wanted &&
+             data.engine_temperature > thresholds.engine_temp_zero &&
+             data.fuel_inj_pressure > thresholds.fuel_pressure_wanted &&
+             data.N2_pressure > thresholds.n2_pressure_zero &&
+             data.LOX_level > thresholds.lox_level_zero &&
+             data.fuel_level > thresholds.fuel_level_zero)
+    {
+        return State::ARMED;
+    }
+    {
+        return State::ARMED;
+    }
+    return State::MANUAL;
+}
+
+State AvState::fromArmed(SensFiltered data, UPLink uplink)
+{
+    if (error())
+    {
+        return State::ERRORGROUND;
+    }
+    else
+    {
+        switch (uplink.id)
+        {
+            case CMD_ID::AV_CMD_IGNITION:
+                if (data.igniter_pressure >= thresholds.igniter_pressure_wanted &&
+                    data.chamber_pressure >= thresholds.chamber_pressure_wanted )
+                {
+                    return State::THRUSTSEQUENCE;
+
+                }else {
+                    return State::ARMED;
+                }
+                break;
+            case CMD_ID::AV_CMD_ABORT:
+                return State::ERRORGROUND;
+            default:
+                State::ARMED;
+        }
+    }
+}
+
+State* possibleStates();
+
+void AvState::update(SensFiltered data, UPLink uplink, bool status)
```

Comment:
Data should be "fed" top the FSM through a separate "registry" (on the wiki I called it the GOAT). Please implement this before merging (no need for thread-safety as of this version).